### ISSUE TITLE
null_resource triggers that interpolate a module output will only work on the 2nd run.

### ISSUE THREAD

**Issue by jonapich** [hashicorp/terraform#15437]
I have a simple module that deploys a lambda function and generates an output using the function's hash. I have confirmed that the hash changes between runs when the zip filename changes, which is the expected behavior.

As a user of the module, whenever the hash changes, I want to push a new configuration. For this I use a local-exec provisioner.

Here's a the gist of the module:
```
resource "aws_lambda_function" "lambda" {
  s3_bucket            = "${var.s3_bucket}"
  s3_key               = "${var.s3_key}"
  (...)
}

output "function_hash" {
  value = "${aws_lambda_function.lamda.source_code_hash}"
}
```

And here's how I am trying to trigger the configuration:
```
module "lambda" { ... }
resource "null_resource" "configuration" {
    triggers { source_code_hash = "${module.lambda.function_hash}" }
    provisioner "local-exec" { ... }
}
```

* On the first run, s3_key ends with `0.1` and the module executes. Then, the null_resource executes.
* I trigger a second run using `0.2` and the module executes. The hash changes. But the null_resource doesn't see the change and doesn't trigger.
* On the third run, still using `0.2`, the lambda doesn't change (as expected) and the null_resource triggers.

It feels to me as if my trigger interpolation happens too early, and uses the value from the tfstate and not the updated value from the module?

Adding a "depends_on" to the null_resource, so that it depends on the module, did not change the behavior.

Is this expected or is this a bug?

**Comment by xoen** [comment]
I'm not sure if I'm affected the same issue:
* I also had to resort to a `null_resource`/`provisioner "local-exec"` to install some 3rd party packages used by my lambda function
* The zip`archive_file` `depends_on` this `local-exec` provisioner to be sure these packages are always in the zip (that was the idea at least!)
* In order for this build script to run when something changes, this local exec provisioner has few `triggers` attributes

I run into a bug which was caused by another dev running `terraform apply` - the build step didn't run as for some reason the triggers did not work.

My educated **guess** is that the old triggers values are stored in the remote store so when someone run applies from a different machine, from terraform perspective they didn't change, therefore the `local-exec` provisioner didn't run.


(It would be lovely to have better handling of lambda functions build/deployments including 3rd party packages installation ❤️. I know this is a limitation of the way AWS Lambda wants functions in a zip but this still feels too complicated for something so common)

**Comment by 2stacks** [comment]
I've come across this same issue and get the following error output.
```
Error applying plan:

1 error(s) occurred:

* null_resource.init-config: null_resource.init-config: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.10.7
    Resource ID: null_resource.init-config
    Mismatch reason: attribute mismatch: triggers.node_ids
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"triggers.%":*terraform.ResourceAttrDiff{Old:"1", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "triggers.node_ids":*terraform.ResourceAttrDiff{Old:"66082801,66082805,66085156", New:"66082801,66082805", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"triggers.%":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.
```
I'm interpolating the resource id's produced by a module and using that as a trigger to rerun a null_provisioner.
```
locals {
  node_ids    = "${module.som-mod.node_id}"
}
resource "null_resource" "init-config" {
  # Changes to any instance of the cluster requires re-provisioning
  triggers {
    node_ids = "${join(",", local.node_ids)}"
  }

  provisioner "local-exec" {
    command = "./scripts/ansible-hosts.sh"
  }
}
```
If I increase the number of nodes added by the module everything works as expected.  If I reduce the number of existing nodes I have to run 'terraform apply' twice to achieve consistency.

**Comment by apparentlymart** [comment]
Hi @jonapich,

I assume that `source_code_hash` in this case isn't being set by you, but rather you're consuming the result generated by the Lambda API after updating your source code.

In that case, I think you're unfortunately running into a current limitation in Terraform where it is unable to predict at plan time that the `source_code_hash` will change during the apply phase, since it does not have enough information to determine this within the configuration alone.

In the short term, the usual workaround for this is to _also_ interpolate into `triggers` a value that changes under _your_ control, from inside the config. From your description, it sounds like the zip file path is a good candidate here.

In the longer term, #14887 creates a new building-block for providers that allows them to tell Terraform core about this sort of "invisible" dependency between attributes. That change alone won't fix this problem, but it _should_ allow someone working on the AWS provider to change the `aws_lambda_function` implementation so that it recognizes that a change to the zip file path (either specified locally or via S3) implies a change to `source_code_hash`, and thus give Terraform core the information it needs to understand that the `null_resource` must be rebuilt in your case.

Since we already have an issue and PR for the Terraform Core change, I'm going to ask our bot to move this issue over to the AWS provider to represent updating the `aws_lambda_function` resource once #14887 is merged and released.

**Comment by hashibot** [comment]
This issue has been automatically migrated to terraform-providers/terraform-provider-aws#1974 because it looks like an issue with that provider. If you believe this is _not_ an issue with the provider, please reply to this issue and let us know.
